File: apps/csharp/design-pattern/src/DesignPattern/Patterns/Factory/Factory.cs
namespace DesignPattern.Patterns.Factory;

// --- Product interfaces ---

public interface IAnimal
{
    string Name { get; }
    string Speak();
    string Eat();
}

public interface IPlant
{
    string Name { get; }
    string Grow();
}

// --- Concrete animals ---

public class Duck : IAnimal
{
    public string Name => "Duck";
    public string Speak() => "Quack!";
    public string Eat() => "Duck eats bugs";
}

public class Frog : IAnimal
{
    public string Name => "Frog";
    public string Speak() => "Croak!";
    public string Eat() => "Frog eats flies";
}

public class Tiger : IAnimal
{
    public string Name => "Tiger";
    public string Speak() => "Roar!";
    public string Eat() => "Tiger eats meat";
}

// --- Concrete plants ---

public class WaterLily : IPlant
{
    public string Name => "WaterLily";
    public string Grow() => "WaterLily grows in water";
}

public class Algae : IPlant
{
    public string Name => "Algae";
    public string Grow() => "Algae spreads in the pond";
}

public class Tree : IPlant
{
    public string Name => "Tree";
    public string Grow() => "Tree grows tall in the jungle";
}

// --- Factory Method pattern (Pond) ---

public abstract class Pond
{
    public IAnimal Animal { get; }
    public IPlant Plant { get; }

    protected Pond()
    {
        Animal = CreateAnimal();
        Plant = CreatePlant();
    }

    protected abstract IAnimal CreateAnimal();
    protected abstract IPlant CreatePlant();

    public string Describe() =>
        $"Pond with {Animal.Name} and {Plant.Name}: {Animal.Speak()}, {Plant.Grow()}";
}

public class DuckPond : Pond
{
    protected override IAnimal CreateAnimal() => new Duck();
    protected override IPlant CreatePlant() => new WaterLily();
}

public class FrogPond : Pond
{
    protected override IAnimal CreateAnimal() => new Frog();
    protected override IPlant CreatePlant() => new Algae();
}

// --- Abstract Factory pattern (OrganismFactory + Habitat) ---

public interface IOrganismFactory
{
    IAnimal CreateAnimal();
    IPlant CreatePlant();
}

public class PondFactory : IOrganismFactory
{
    public IAnimal CreateAnimal() => new Frog();
    public IPlant CreatePlant() => new Algae();
}

public class JungleFactory : IOrganismFactory
{
    public IAnimal CreateAnimal() => new Tiger();
    public IPlant CreatePlant() => new Tree();
}

public class Habitat
{
    public IAnimal Animal { get; }
    public IPlant Plant { get; }

    public Habitat(IOrganismFactory factory)
    {
        Animal = factory.CreateAnimal();
        Plant = factory.CreatePlant();
    }

    public string Describe() =>
        $"Habitat with {Animal.Name} and {Plant.Name}: {Animal.Speak()}, {Plant.Grow()}";
}

File: apps/ruby/design-pattern/14-factory/lib/factory.rb
# frozen_string_literal: true

# === 動物クラス ===

class Duck
  def speak
    "Quack!"
  end

  def eat
    "Duck is eating."
  end

  def sleep
    "Duck is sleeping."
  end
end

class Frog
  def speak
    "Croak!"
  end

  def eat
    "Frog is eating."
  end

  def sleep
    "Frog is sleeping."
  end
end

class Tiger
  def speak
    "Roar!"
  end

  def eat
    "Tiger is eating."
  end

  def sleep
    "Tiger is sleeping."
  end
end

# === 植物クラス ===

class WaterLily
  def grow
    "WaterLily is growing."
  end
end

class Algae
  def grow
    "Algae is growing."
  end
end

class Tree
  def grow
    "Tree is growing."
  end
end

# === Factory Method パターン ===

# 池の基底クラス
class Pond
  def initialize(number_animals, animal_class, number_plants, plant_class)
    @animals = Array.new(number_animals) { animal_class.new }
    @plants = Array.new(number_plants) { plant_class.new }
  end

  def simulate_one_day
    output = []
    @plants.each { |plant| output << plant.grow }
    @animals.each do |animal|
      output << animal.speak
      output << animal.eat
      output << animal.sleep
    end
    output
  end
end

# アヒルの池（Factory Method）
class DuckPond < Pond
  def initialize(number_animals: 3, number_plants: 2)
    super(number_animals, Duck, number_plants, WaterLily)
  end
end

# カエルの池（Factory Method）
class FrogPond < Pond
  def initialize(number_animals: 3, number_plants: 2)
    super(number_animals, Frog, number_plants, Algae)
  end
end

# === Abstract Factory パターン ===

# 生物工場
class OrganismFactory
  def initialize(plant_class, animal_class)
    @plant_class = plant_class
    @animal_class = animal_class
  end

  def new_animal
    @animal_class.new
  end

  def new_plant
    @plant_class.new
  end
end

# 生息地（Abstract Factory を使用）
class Habitat
  def initialize(number_animals, number_plants, organism_factory)
    @animals = Array.new(number_animals) { organism_factory.new_animal }
    @plants = Array.new(number_plants) { organism_factory.new_plant }
  end

  def simulate_one_day
    output = []
    @plants.each { |plant| output << plant.grow }
    @animals.each do |animal|
      output << animal.speak
      output << animal.eat
      output << animal.sleep
    end
    output
  end
end
